src/classes/ADDR_Contact_TDTM.cls

```diff
@@ -188,9 +200,8 @@ public class ADDR_Contact_TDTM extends TDTM_Runnable {
                     continue;
                 }
 
-                // if the address changed, remember the contact we want to add a new address for    
-                Contact conOld = (listOld != null ? (Contact)listOld[i] : null);
-                if (isContactAddressChanged(con, conOld)) 
+                // if the contacts address changed or if the contact with an address moves to a new household remember the contact
+                if (isContactAddressChanged(con, conOld) || (acctChanged && !isContactAddressEmpty(con))) 
```

Comment:
Yes, but I think it is ok, because if the address is changed in the before context then it will check and deal with duplicates but if it is in the after context and not a duplicate it will add it to the address. I think this is the trickery you mentioned in your comment.